cpp/src/strings/regex/utilities.cuh

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <strings/regex/regex.cuh>
+
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/get_value.cuh>
+#include <cudf/detail/utilities/cuda.cuh>
+#include <cudf/strings/detail/utilities.cuh>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/scan.h>
+
+namespace cudf {
+namespace strings {
+namespace detail {
+
+template <typename ForEachFunction>
+__global__ void for_each_kernel(ForEachFunction fn, reprog_device const d_prog, size_type size)
```

Comment:
These names are very generic, but they are fairly specific in use. Is there a chance that another `strings::detail` kernel would be named similarly? Maybe `for_each_re_kernel`?

Reply:
It is certainly possibly for the name to be reused but not with these parameters. I felt the `reprog_device` parameter specifically communicated this was to be used with regex kernels.

Reply:
I suspect that you're right and that while this is coming up during review (for both Mike and me) it will probably be clear to readers in the future.

Reply:
I'll have to disagree with this one. Will this prevent approval of this PR?